### ISSUE TITLE
FRAME: #[run_only_once] for top-level migration

### ISSUE THREAD

**Issue by ordian** [paritytech/polkadot-sdk#2421]
### Motivation

Currently, we have [`VersionedMigration`](https://github.com/paritytech/polkadot-sdk/blob/552be4800d9e4b480f79a300fc531783e04be099/substrate/frame/support/src/migrations.rs#L70) for migrating pallets, which protects us from running the same migration twice. However, we don't have a similar tool for top-level `on_runtime_upgrade` migrations, e.g. for migrating session keys, since there's no `StorageVersion`. We got bitten by this already: https://github.com/paritytech/polkadot-sdk/issues/1149#issuecomment-1811381184

### Solution

One potential solution could be using a storage item and a unique name of a migration, e.g.
#[run_only_once(RemoveImOnlineSessionKey)]. That would write into a storage item along with the current spec version. And some autocleanup based on spec versions. Instead of using spec version, might be better to have a runtime upgrade counter. This way we can do autocleanup e.g. after 10 runtime upgrades. Runtime spec version OTOH can bump unpredictably in the future. 

Bikeshed on naming, syntax and suggestions for alternative solutions are welcome :)

**Comment by ggwpez** [comment]
>Currently, we have `VersionedMigration` for migrating pallets

Yea I thought about non-pallet use cases when we designed it. AFAIK it should be possibly to use for general applications, we just need to implement it for something else than `GetStorageVersion + PalletInfoAccess` but instead a more generic version getter that can read directly from storage.  
Could be done with some adapter types that then wrap a pallet... cc @liamaharon

**Comment by liamaharon** [comment]
Can we just gate them with `if spec_version == intended_spec_version_to_run { do_migration }`? 
 
Note this will be mandatory when using new versions of `try-runtime-cli` which enforce that migrations are idempotent.

**Comment by liamaharon** [comment]
Allow me to play devils advocate for a minute :P 

What are some concrete cases where a migration cannot be written to be idempotent and requires some external handling like is proposed here? Could the session key logic really not have been written to be idempotent?

A significant benefit of writing migrations to be idempotent at their core is that they are **always safe**. Otherwise, they are **unsafe by default** and depends on every developer using the migration API to 1. be aware that the migration is unsafe and 2. know how to use the new abstraction correctly and 3. remember to use the abstraction.

**Comment by liamaharon** [comment]
In the case of the session keys I think at its root we'd want `Session::upgrade_keys` to made idempotent, not the migration at the level above right?

**Comment by kianenigma** [comment]
..or just deprecate `Hooks::on_runtime_upgrade`. Anything that can be written `Hooks::on_runtime_upgrade` can directly be written as `OnRuntimeUpgrade`. 

Some pallets like XCM and pallet-contracts have complicated internal migrations that are better written with `Hooks::on_runtime_upgrade`. If you write them as `OnRuntimeUpgrade` you need a way to assert that they are always included in the runtime. 

If we pursue the goal of removing migrations from `type Executive` and moving them to `frame_system`, which would arguably make more sense for other reasons as well, then a pallet can assert in `integrity_check` that a certain migration is present in the runtime. 

Then, a pallet like `frame_system` can assert that all of its important signed extensions are part of the runtime. 

A pallet like XCM can assert that its omni-present migration script is always part of the runtime.

**Comment by bkchr** [comment]
> ..or just deprecate `Hooks::on_runtime_upgrade`. Anything that can be written `Hooks::on_runtime_upgrade` can directly be written as `OnRuntimeUpgrade`.

This was not the issue here. I mean we discourage any kind of usage of migrations in the pallets.



> Then, a pallet like `frame_system` can assert that all of its important signed extensions are part of the runtime.
> 
> A pallet like XCM can assert that its omni-present migration script is always part of the runtime.

This is only something we don't want to encourage. Migrations should not always be included and be removed when they are not required anymore.

**Comment by kianenigma** [comment]
> This is only something we don't want to encourage. Migrations should not always be included and be remoaved when they are not required anymore.

Yes, but afaik there are exceptions to this, such as pallet-xcm and pallet-contracts which at least at some point had correct and sound migrations that were designed to be executed all the time, and I would not say there are a bad design. Complicated migrations probably need to be like this and we should allow and faciliatte it, but 90% of migrations should be on-off.
